File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
    stack<int>s1;
    int minEle = INT_MAX;
    MinStack() {
        
    }
    
    void push(int val) {
        if(val <= minEle) {
            s1.push(minEle);
            minEle = val;
        }
        s1.push(val);
    }
    
    void pop() {
        if(s1.top() == minEle) {
            s1.pop();
            minEle = s1.top();
        }
        s1.pop();
    }
    
    int top() {
        return s1.top();
    }
    
    int getMin() {
        return minEle;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.java
class MinStack {
    Stack<Integer>st = new Stack();
    int minEle = Integer.MAX_VALUE;
    public MinStack() {
        
    }
    
    public void push(int val) {
        if(val <= minEle) {
            st.push(minEle);
            minEle = val;
        }
        st.push(val);
    }
    
    public void pop() {
        if(minEle == st.peek()) {
            st.pop();
            minEle = st.peek();
        }
        st.pop();
    }
    
    public int top() {
        return st.peek();
    }
    
    public int getMin() {
        return minEle;
    }
}

/*

class MinStack {
    Stack<Integer>st = new Stack();
    Stack<Integer>min = new Stack();
    public MinStack() {
        
    }
    
    public void push(int val) {
        if(st.isEmpty()) {
            st.push(val);
            min.push(val);
        }
        else {
            st.push(val);
            min.push(Math.min(min.peek(), val));
        }
    }
    
    public void pop() {
        st.pop();
        min.pop();
    }
    
    public int top() {
        return st.peek();
    }
    
    public int getMin() {
        return min.peek();
    }
}


*/

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */